charmhelpers/contrib/network/ovs/ovsdb.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2019 Canonical Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import uuid
+
+from . import utils
+
+
+class SimpleOVSDB(object):
+    """Simple interface to OVSDB through the use of command line tools.
+
+    OVS and OVN is managed through a set of databases.  These databases have
+    similar command line tools to manage them.  We make use of the similarity
+    to provide a generic class that can be used to manage them.
+
+    The OpenvSwitch project does provide a Python API, but on the surface it
+    appears to be a bit too involved for our simple use case.
+
+    Examples:
+    chassis = SimpleOVSDB('ovn-sbctl', 'chassis')
+    for chs in chassis:
+        print(chs)
+
+    bridges = SimpleOVSDB('ovs-vsctl', 'bridge')
+    for br in bridges:
+        if br['name'] == 'br-test':
+            bridges.set(br['uuid'], 'external_ids:charm', 'managed')
+    """
+
+    def __init__(self, tool, table):
+        """SimpleOVSDB constructor
+
+        :param tool: Which tool with database commands to operate on.
+                     Usually one of `ovs-vsctl`, `ovn-nbctl`, `ovn-sbctl`
+        :type tool: str
+        :param table: Which table to operate on
+        :type table: str
+        """
+        self.tool = tool
```

Comment:
It might be useful to validate the `tool` really is one of the tools that can be used.  i.e.

```python
if tool not in ['ovs-vsctl', 'ovn-nbctl', 'ovn-sbctl']:
    raise RuntimeError("param tool passed to SimpleOVSCB() not one of ...")
```

Reply:
Since most all OVS/OVN command line tools that interface to a underlying OVSDB Server support the command line interfaces used the ``SimpleOVSDB`` class may be useful for tools we do not know about or have yet envisioned.  So I think validation of the specific tool used here may be limiting the user.  We could of course log/print a warning if used with something we currently don't know about?

Reply:
It's a library function; we should always verify that it's the things that we allow rather than being permissive -- otherwise bugs (like getting the wrong tool name) propagate to weird places in the code.

Reply:
Yes, you are right.